crates/adapters/src/samply.rs

```diff
@@ -0,0 +1,271 @@
+/// executing.
+///
+/// When the guard is dropped, the span will be exited.
+///
+/// This is returned by [SamplySpan::entered].
+pub struct EnteredSamplySpan {
+    span: EnteredSpan,
+    start: Option<Timestamp>,
+}
+
+impl EnteredSamplySpan {
+    fn new(span: SamplySpan) -> Self {
+        Self {
+            start: if span.span.is_disabled() || !enabled!(Level::DEBUG) {
+                None
+            } else {
+                Some(Timestamp::now())
+            },
+            span: span.span.entered(),
+        }
+    }
+
+    /// Exits the span, returning the underlying [SamplySpan].
+    #[allow(dead_code)]
+    pub fn exit(mut self) -> SamplySpan {
+        self.do_exit();
+        SamplySpan {
+            span: std::mem::replace(&mut self.span, Span::none().entered()).exit(),
+        }
+    }
+
+    fn do_exit(&mut self) {
+        if let Some(start) = self.start.take() {
+            write_marker(
+                start,
+                Timestamp::now(),
+                self.span.metadata().unwrap().name(),
+            );
+        }
+    }
+}
+
+impl Drop for EnteredSamplySpan {
+    fn drop(&mut self) {
+        self.do_exit();
+    }
+}
+
+fn markers_dir() -> Option<&'static Path> {
+    static MARKERS_DIR: LazyLock<Option<PathBuf>> = LazyLock::new(|| {
+        tempdir()
+            .map(TempDir::keep)
+            .inspect_err(|e| error!("Failed to create temporary directory: {e}"))
+            .ok()
+    });
+
+    Some(LazyLock::force(&MARKERS_DIR).as_ref()?)
+}
+
+fn with_marker_file<F, R>(f: F) -> Option<R>
+where
+    F: FnOnce(&mut File) -> R,
+{
+    fn create_marker_file() -> Option<File> {
+        let file_name = markers_dir()?.join(format!(
+            "marker-{}-{}.txt",
+            nix::unistd::getpid(),
+            nix::unistd::gettid()
+        ));
+
+        // Create the file.
+        let file = File::options()
+            .create(true)
+            .write(true)
+            .read({
+                // We aren't going to read from the file, so we ordinarily
+                // wouldn't need read permission, but we're about mmap it and
+                // mmap always requires read permission.
+                true
+            })
+            .open(&file_name)
+            .inspect_err(|e| error!("Failed to create marker file {e}: {}", file_name.display()))
+            .ok()?;
+
+        // Create an mmap for the file.  This cannot be skipped because it
+        // triggers samply to read and interpret the file.  We will not use it
+        // to write to the file (because writing to a text file via mmap is
+        // painful and we haven't yet proved that it is a performance problem),
+        // so it is not necessary to map it with any particular protection or
+        // flags, so we use PROT_READ because that offers the fewest ways to
+        // screw up.
+        unsafe {
+            nix::sys::mman::mmap(
+                None,
+                NonZeroUsize::new(4096).unwrap(),
```

Comment:
How do you know this is enough?

Reply:
The amount literally doesn't matter, because we aren't using the mapping to read or write and `samply` reads the whole file regardless of how much we mmap. `samply` just uses the fact that the mapping exists as a signal to read the file.

I'll add a comment.

Come to think of it, I should get the page size instead of hard-coding 4096 (I think that the ARM page size is sometimes 16k or 64k).